health check: structured active healthcheck logging

jsedgwick opened the pull request:
*Description*:
As discussed in #2028 

The idea is to transition to proto logging when real sinks are available, but right now I stop short and dump the proto to a file in JSON form. Compare with the manual JSON of outlier detection logging, which could now be migrated to this approach.

The only TODO is docs. Where should this be documented considering there are proto changes as well?

*Risk Level*: Low / Medium

*Testing*:
Added new tests and updated existing

*Docs Changes*:
TBD pending discussion of approach. Where should I document this logging? 

*Release Notes*:
Will add w/ docs.

Fixes #2028

mattklein123 commented:
@junr03 please take first pass.

junr03 commented on source/common/upstream/health_checker_impl.cc:
```diff
@@ -23,37 +23,48 @@ class HealthCheckerFactoryContextImpl : public Server::Configuration::HealthChec
+  HealthCheckEventLoggerSharedPtr event_logger;
+  if (!hc_config.event_log_path().empty()) {
+    event_logger =
+        std::make_shared<HealthCheckEventLoggerImpl>(log_manager, hc_config.event_log_path());
```
Two questions about the event logger's shared_ptr:

1. ~It seems that each health checker is the sole owner of the event logger, as each health checker construction is preceded by a event logger construction. Is the reason to make is a shared ptr because below we pass the event logger to a factory context and that context can share the event logger to more owners?~ Ah I see, the ownership is shared between the health checker and all the active sessions. Why wouldn't the healthchecker own the event logger and give a reference to the active sessions? Or have the active session write to the `parent_`'s `event_logger_`?

2. Why do we return the event logger's shared ptr from the context, and pass it into constructors by reference?

jsedgwick replied:
Gosh, I can't remember, I kind of just fixed a couple signatures and then propagated necessary changes until everything compiled. But you're right that it all feels weird. I'll refactor so it makes more sense. I think I only had shared_ptrs in the first place because that's how the outlier loggers were structured..

junr03 commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,53 @@
+message ActiveHealthCheckEvent {
+  string host_address = 1 [(validate.rules).string.min_bytes = 1];
+  string cluster_name = 2 [(validate.rules).string.min_bytes = 1];
+
```
can we log the type of health checking that caused the event?

jsedgwick commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,53 @@
+// [#protodoc-title: Health check logging]
+// [#proto-status: draft]
+
+message ActiveHealthCheckEvent {
```
Yeah I didn't know when/if we'd get around to converting the outlier logging to proto, and I liked the idea of colocating the associated component's proto w/ the logging proto, so I just kept it simple for now. I left the proto as draft status so we can always refactor in the future.

junr03 commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,62 @@
+  }
+}
+
+enum HealthCheckFailureType {
```
thanks for adding this. I was also wondering about adding the type of health check that failed (redis, http, tcp).

jsedgwick replied:
Oh! Right. Sure, will do.

jsedgwick commented:
PTAL @junr03

jsedgwick commented:
PTAL @danielhochman

danielhochman commented:
needs docs. is there still some question about where to put them? i would think a section in the arch overview health checking docs (similar to outlier detection logging). and a reference in [common_messages.rst](https://github.com/envoyproxy/envoy/blob/master/docs/root/api-v2/common_messages/common_messages.rst) 

i would also add a TODO or open an issue to convert outlier detection to use the same logging mechanism.

jsedgwick commented:
PTAL @junr03 @danielhochman

jsedgwick commented:
also opened #3405

junr03 commented on docs/root/intro/arch_overview/health_checking.rst:
```diff
@@ -24,6 +24,15 @@ unhealthy, successes required before marking a host healthy, etc.):
+--------------------------
+
+A per-healthchecker log of ejection and addition events can optionally be produced by Envoy by
+specifying a log file path in `the HealthCheckConfig <envoy_api_field_core.HealthCheck.event_log_path>`.
```
nit: use snake_case for field names

mattklein123 commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,71 @@
+}
+
+message HealthCheckEjectUnhealthy {
+  // The type of failure that caused this ejection
```
nit: end sentences with full stops (same in other places).

mattklein123 commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,71 @@
+  HealthCheckFailureType failure_type = 1;
+
+  // The timeout after which health checks fail for hosts in this cluster
+  google.protobuf.Duration timeout = 2 [(validate.rules).duration.required = true];
```
It's a little strange to include timeout when we also log Active/Passive failures. Additionally, we aren't logging all the other types of timeouts that have been added. Unless we are going to log times specific to this event, I would probably drop timeout/unhealthy_threshold from this log since it can be intuited from the config.

mattklein123 commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,71 @@
+  // The number of healthy health checks required before a host is marked
+  // healthy. Note that during startup, only a single successful health check is
+  // required to mark a host healthy.
+  google.protobuf.UInt32Value healthy_threshold = 1;
```
for similar reasons as above I would drop this.

mattklein123 commented on api/envoy/api/v2/core/health_check_logging.proto:
```diff
@@ -0,0 +1,71 @@
+
+  // Whether this addition is the result of the first ever health check on a host, in which case
+  // the above healthy_threshold is bypassed and the host is immediately added.
+  bool first_check = 2;
```
Fine with me if you want to keep this since it is actually situationally dependent, but I could go either way.

mattklein123 commented on source/common/protobuf/utility.cc:
```diff
@@ -87,7 +87,8 @@ void MessageUtil::loadFromFile(const std::string& path, Protobuf::Message& messa
 std::string MessageUtil::getJsonStringFromMessage(const Protobuf::Message& message,
-                                                  const bool pretty_print) {
+                                                  const bool pretty_print,
+                                                  const bool always_print_primitive_fields) {
```
humorously, I was just looking at adding this in a different change. Cool!

mattklein123 commented on source/common/protobuf/utility.h:
```diff
@@ -213,7 +213,8 @@ class MessageUtil {
   static std::string getJsonStringFromMessage(const Protobuf::Message& message,
-                                              bool pretty_print = false);
+                                              bool pretty_print = false,
+                                              bool always_print_primitive_fields = false);
```
nit: update doc comment

mattklein123 commented:
Oh, please add a release note also. Thank you!

mattklein123 commented:
@jsedgwick see https://github.com/envoyproxy/envoy/pull/3494. Let's put the health check access log proto somewhere in the "output" folder (or whatever we end up calling it when that PR is merged). Thank you! cc @htuch

mattklein123 commented on api/envoy/api/v2/core/health_check.proto:
```diff
@@ -176,6 +176,9 @@ message HealthCheck {
   // The default value for "healthy edge interval" is the same as the default interval.
   google.protobuf.Duration healthy_edge_interval = 16;
+
+  // Specifies the path to the health check event log.
```
Can we cross link to relevant docs here about the event log, and also specify that if empty, no event log will be written?

mattklein123 commented on api/envoy/data/core/v2alpha/health_check_event.proto:
```diff
@@ -0,0 +1,57 @@
+// :ref:`Health check logging <arch_overview_health_check_logging>`.
+
+message HealthCheckEvent {
+  HealthCheckerType health_checker_type = 1;
```
can we add enum validation here

mattklein123 commented on api/envoy/data/core/v2alpha/health_check_event.proto:
```diff
@@ -0,0 +1,57 @@
+
+message HealthCheckEjectUnhealthy {
+  // The type of failure that caused this ejection.
+  HealthCheckFailureType failure_type = 1;
```
enum validation

mattklein123 commented on api/envoy/data/core/v2alpha/health_check_event.proto:
```diff
@@ -0,0 +1,57 @@
+
+message HealthCheckEvent {
+  HealthCheckerType health_checker_type = 1;
+  string host_address = 2 [(validate.rules).string.min_bytes = 1];
```
@htuch @mrice32 I haven't been fully tracking the convo in https://github.com/envoyproxy/envoy/pull/3478. What are we thinking here for this type of data? Should we be using the full address type?

htuch replied:
Yes, full address preferable. The only situation it might not be the preferred choice is if we don't have the possibility of there being a port encoded (i.e. not `:80`), it's definitely TCP and it can't be a pipe..

htuch replied:
One thing from #3478... what happens when we have hosts that appear at multiple levels and priorities? Do we have unique HC events, do we have locality/priority information associated with them?

htuch commented on api/envoy/data/core/v2alpha/health_check_event.proto:
```diff
@@ -0,0 +1,57 @@
+
+option (gogoproto.equal_all) = true;
+
+// [#protodoc-title: Health check logging events]
```
If this is logging, doesn't it belong in `envoy.data`?

mattklein123 replied:
It is in `envoy.data`?

htuch replied:
My bad, I was focusing on the `core` :)

htuch commented on api/envoy/data/core/v2alpha/health_check_event.proto:
```diff
@@ -0,0 +1,57 @@
+enum HealthCheckFailureType {
+  ACTIVE = 0;
+  PASSIVE = 1;
+  NETWORK = 2;
```
Please also consider the discussion in https://github.com/envoyproxy/envoy/pull/3478#discussion_r192235554

jsedgwick commented:
Sorry folks, this flew off my radar - I updated per above comments with the exception of 
https://github.com/envoyproxy/envoy/pull/3176#discussion_r192856278
since I couldn't tell if there was a conclusion and whether the enum I introduced is inappropriate. As Matt pointed out its v2alpha so probably doesn't have to settles now.

PTAL @htuch @mattklein123 @danielhochman. Thanks!

mattklein123 commented on api/envoy/api/v2/core/health_check.proto:
```diff
@@ -185,6 +185,10 @@ message HealthCheck {
   // The default value for "healthy edge interval" is the same as the default interval.
   google.protobuf.Duration healthy_edge_interval = 16;
+
+  // Specifies the path to the :ref:`healthy check event log <arch_overview_health_check_logging>`.
```
s/healthy/health

mattklein123 commented on api/envoy/api/v2/core/health_check.proto:
```diff
@@ -185,6 +185,10 @@ message HealthCheck {
   google.protobuf.Duration healthy_edge_interval = 16;
+
+  // Specifies the path to the :ref:`healthy check event log <arch_overview_health_check_logging>`.
+  // health check event log. If empty, no event log will be written.
```
Remove duplicate "health check event log."

mattklein123 commented on source/common/upstream/health_checker_base_impl.h:
```diff
@@ -127,5 +128,22 @@ class HealthCheckerImplBase : public HealthChecker,
+  HealthCheckEventLoggerImpl(AccessLog::AccessLogManager& log_manager, const std::string& file_name)
+      : file_(log_manager.createAccessLog(file_name)) {}
+
+  virtual ~HealthCheckEventLoggerImpl() {}
```
nit: delete

jsedgwick commented:
(still working on the DCO issue)

jsedgwick commented:
oops, will fix docs
